### PR TITLE
Fixes Partial web interface  #231

### PR DESCRIPTION
**Pull request by tobi01001**
Script function to load css/js from the ESP when offline / error.

### CODE REVIEW

**Comment by henrygab** [comment]
I also put my devices onto a separate (non-internet-connected) network, so I'm interested in this change.  My javascript skills are not great, so while I've looked at the PR, my review is likely insufficient.

At the same time, I do wonder ...

@jasoncoon -- Can you help me understand the benefits to obtaining the files via the internet, vs. always serving them from the device?

I ask because the reasons I immediately think of don't appear to have much impact:

1. Bugfixes / security updates:  But, specific versions are requested, so it seems the CDN versions would still give the specific versions requested?
2. License related distribution concerns:  But, each of the libraries appear to be MIT-licensed, and the code is already in the device's local file system by default?
3. Less wireless traffic to ESP:  This could be... although if the ESP provides the HTTP headers to indicate these files can be cached, web clients will only request them from the device a single time.

<details><summary>Apparent library list & direct links to corresponding licenses</summary><P/>

* [jquery @ 3.6.0](https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js) -- [MIT](https://jquery.org/license/) license
* [jquery.minicolors @ 2.3.6](https://cdnjs.cloudflare.com/ajax/libs/jquery-minicolors/2.3.6/jquery.minicolors.min.css) -- [MIT](https://github.com/claviska/jquery-minicolors/blob/master/LICENSE.md) license
* [bootstrap @ 5.1.3](https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css) -- [MIT](https://getbootstrap.com/docs/5.1/about/license/) license
* [bootstrap icons @ 1.7.2](https://cdn.jsdelivr.net/npm/bootstrap-icons@1.7.2/font/bootstrap-icons.min.css) -- [MIT](https://github.com/twbs/icons/blob/main/LICENSE.md) license
* [FileSaver @ 2.0.5](https://cdnjs.cloudflare.com/ajax/libs/FileSaver.js/2.0.5/FileSaver.min.js) -- [MIT](https://github.com/eligrey/FileSaver.js/blob/master/LICENSE.md) license
* ~~[reconnecting-websocket @ 1.0.0](https://cdnjs.cloudflare.com/ajax/libs/reconnecting-websocket/1.0.0/reconnecting-websocket.min.js)~~ -- [MIT](https://github.com/joewalnes/reconnecting-websocket/blob/master/LICENSE.txt) license, even if still commented out in current code)

</details>

File: esp8266-fastled-webserver/data/index.htm [hunk 1]
```diff
@@ -7,12 +7,45 @@
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <title>ESP8266 + FastLED</title>
 
-  <!-- request CSS from internet CDN -->
-  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
-  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery-minicolors/2.3.6/jquery.minicolors.min.css" integrity="sha512-BVeRnUOL0G7d4gXmj+0VxpoiQuEibKQtlkclADKvCdNrESs0LA6+H8s1lU455VqWFtHBfF/pKDGw/CMat2hqOg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
-  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons@1.7.2/font/bootstrap-icons.min.css">
+  <!-- Fallback to load resources from FileSystem if not available online -->
```
Comment (henrygab) [inline esp8266-fastled-webserver/data/index.htm:10]: Question:

Would it make sense to have a common `.js` file, which has the common script functions?

File: esp8266-fastled-webserver/data/index.htm [hunk 2]
```diff
@@ -7,12 +7,45 @@
...
+  <script>
+    function CssJsFallback(el) {
+      if (el instanceof HTMLScriptElement || el instanceof HTMLLinkElement) {
+        let Url = el.getAttribute('data-fallback');
+        if (Url) {
+          if(el instanceof HTMLScriptElement)
+          {
+            document.write('<script src="'+Url+'">\x3C/script>');
```
Comment (henrygab) [inline esp8266-fastled-webserver/data/index.htm:18]: Why the `\x3C` here?  0x3C == `<`, which is included as a literal earlier in the same string?

File: esp8266-fastled-webserver/data/index.htm [hunk 3]
```diff
@@ -7,12 +7,45 @@
...
-  <!-- request CSS from internet CDN -->
-  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
-  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery-minicolors/2.3.6/jquery.minicolors.min.css" integrity="sha512-BVeRnUOL0G7d4gXmj+0VxpoiQuEibKQtlkclADKvCdNrESs0LA6+H8s1lU455VqWFtHBfF/pKDGw/CMat2hqOg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
-  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons@1.7.2/font/bootstrap-icons.min.css">
+  <!-- Fallback to load resources from FileSystem if not available online -->
+  <script>
+    function CssJsFallback(el) {
+      if (el instanceof HTMLScriptElement || el instanceof HTMLLinkElement) {
+        let Url = el.getAttribute('data-fallback');
+        if (Url) {
```
Comment (henrygab) [inline esp8266-fastled-webserver/data/index.htm:15]: This appears to modify the attributes of the element that fails to load via use of the `onerror()` callback.

Question:
Is there a specification that requires the element to be re-loaded when its attributes are changed in this callback?

File: esp8266-fastled-webserver/data/index.htm [hunk 4]
```diff
@@ -7,12 +7,45 @@
...
+          }
+        }
+        // No integrity check for local resource
+        let integrity = el.getAttribute('integrity');
+        if(integrity) {
+          el.removeAttribute('integrity');
+        }
+      }
+    }
+  </script>
```
Comment (henrygab) [inline esp8266-fastled-webserver/data/index.htm:32]: This appears to modify the attributes of the element that fails to load via use of the `onerror()` callback, and for this question, I will presume that modification of the element will force a web browser to attempt to reload the script.

Question:
If there is a problem loading the file from the original URI, what happens if the attempt to load from the local file system also fails?

Will this go into an infinite loop?

If so, would it be appropriate to change or clear the `onerror` attribute here?

File: esp8266-fastled-webserver/data/index.htm (hunk 2 above)
Comment (tobi01001) [reply esp8266-fastled-webserver/data/index.htm:18]: Good question. I tried without (putting the literal directly) which did not work. So it seems to be there for a reason (I do also not know currently).

File: esp8266-fastled-webserver/data/index.htm (hunk 3 above)
Comment (tobi01001) [reply esp8266-fastled-webserver/data/index.htm:15]: [https://www.w3schools.com/jsref/event_onerror.asp](https://www.w3schools.com/jsref/event_onerror.asp) is the specification for `onerror()` compatible to "all" common browsers.

There is no direct reload neccessary. At least in chrome it just reloads the ressource from the fallback path.

File: esp8266-fastled-webserver/data/index.htm (hunk 4 above)
Comment (tobi01001) [reply esp8266-fastled-webserver/data/index.htm:32]: On the second attempt (the fallback) it will try once more but then the fallback would be equal to the current element.
If it works, it looksmlike this:
image

If both are wrong it just stops after the second one (maybe as there is nothing to be replaced being the same path):
image

**Comment by tobi01001** [comment]
> I also put my devices onto a separate (non-internet-connected) network, so I'm interested in this change. My javascript skills are not great, so while I've looked at the PR, my review is likely insufficient.
> 
> At the same time, I do wonder ...
> 
> @jasoncoon -- Can you help me understand the benefits to obtaining the files via the internet, vs. always serving them from the device?
> 
> I ask because the reasons I immediately think of don't appear to have much impact:
> 
> 1. Bugfixes / security updates:  But, specific versions are requested, so it seems the CDN versions would still give the specific versions requested?
> 2. License related distribution concerns:  But, each of the libraries appear to be MIT-licensed, and the code is already in the device's local file system by default?
> 3. Less wireless traffic to ESP:  This could be... although if the ESP provides the HTTP headers to indicate these files can be cached, web clients will only request them from the device a single time.
> 
> Apparent library list & direct links to corresponding licenses
> * [jquery @ 3.6.0](https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js) -- [MIT](https://jquery.org/license/) license
> * [jquery.minicolors @ 2.3.6](https://cdnjs.cloudflare.com/ajax/libs/jquery-minicolors/2.3.6/jquery.minicolors.min.css) -- [MIT](https://github.com/claviska/jquery-minicolors/blob/master/LICENSE.md) license
> * [bootstrap @ 5.1.3](https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css) -- [MIT](https://getbootstrap.com/docs/5.1/about/license/) license
> * [bootstrap icons @ 1.7.2](https://cdn.jsdelivr.net/npm/bootstrap-icons@1.7.2/font/bootstrap-icons.min.css) -- [MIT](https://github.com/twbs/icons/blob/main/LICENSE.md) license
> * [FileSaver @ 2.0.5](https://cdnjs.cloudflare.com/ajax/libs/FileSaver.js/2.0.5/FileSaver.min.js) -- [MIT](https://github.com/eligrey/FileSaver.js/blob/master/LICENSE.md) license
> * ~[reconnecting-websocket @ 1.0.0](https://cdnjs.cloudflare.com/ajax/libs/reconnecting-websocket/1.0.0/reconnecting-websocket.min.js)~ -- [MIT](https://github.com/joewalnes/reconnecting-websocket/blob/master/LICENSE.txt) license, even if still commented out in current code)

On my side, the reason is performance (so more or less 3) . The ESP does as little as neccessary (still with room for improvement though). The browser loads the stylesheets and script ressources directly from the web...
Though, when chached the impact might not be huge.

File: esp8266-fastled-webserver/data/index.htm (hunk 2 above)
Comment (jasoncoon) [reply esp8266-fastled-webserver/data/index.htm:18]: https://stackoverflow.com/a/236106

**Comment by jasoncoon** [comment]
I preferred online CDN hosted resources (css, js, etc) over local ESP hosted resources initially because the ESP webserver had problems serving multiple concurrent requests. The browser would load the html and then immediately and concurrently request all of the other resources from the ESP. Sometimes the requests would hang indefinitely, others the ESP would crash.  This doesn't seem to be as much of a problem with newer library versions.  With client-side caching, it's not a problem after the first reload after any changes are made.

This PR seems like a great solution, I've just needed to find time to test it.  If you (@tobi01001 and/or @henrygab) have already tested it, both on and offline, I trust you and we can merge it.  :)

File: esp8266-fastled-webserver/data/index.htm (hunk 2 above)
Comment (henrygab) [reply esp8266-fastled-webserver/data/index.htm:18]: thanks for finding the answer, Jason!  The edge cases are always interesting.

**Comment by henrygab** [comment]
> This PR seems like a great solution, I've just needed to find time to test it. If you (@tobi01001 and/or @henrygab) have already tested it, both on and offline, I trust you and we can merge it. :)

I have limited time these for at least the 1-2 weeks.  I can only say it "looks" correct.  I don't know how it acts for different browsers, although @tobi01001 shows results under chrome in both online and offline cases.
